File: app/src/main/java/com/peterlaurence/trekadvisor/service/LocationService.java
package com.peterlaurence.trekadvisor.service;

import android.Manifest;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.location.Location;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.support.v4.app.ActivityCompat;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.peterlaurence.trekadvisor.MainActivity;
import com.peterlaurence.trekadvisor.R;
import com.peterlaurence.trekadvisor.core.TrekAdvisorContext;
import com.peterlaurence.trekadvisor.menu.events.RecordGpxStartEvent;
import com.peterlaurence.trekadvisor.menu.events.RecordGpxStopEvent;
import com.peterlaurence.trekadvisor.service.event.GpxFileWriteEvent;
import com.peterlaurence.trekadvisor.service.event.LocationServiceStatus;
import com.peterlaurence.trekadvisor.util.gpx.GPXWriter;
import com.peterlaurence.trekadvisor.util.gpx.model.Gpx;
import com.peterlaurence.trekadvisor.util.gpx.model.Track;
import com.peterlaurence.trekadvisor.util.gpx.model.TrackPoint;
import com.peterlaurence.trekadvisor.util.gpx.model.TrackSegment;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.io.File;
import java.io.FileOutputStream;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class LocationService extends Service {
    private static final String GPX_VERSION = "1.1";
    private static final String NOTIFICATION_ID = "peterlaurence.LocationService";
    private static final int SERVICE_ID = 126585;
    private Looper mServiceLooper;
    private Handler mServiceHandler;
    private FusedLocationProviderClient mFusedLocationClient;
    private LocationRequest mLocationRequest;
    private LocationCallback mLocationCallback;

    private List<TrackPoint> mTrackPoints;

    private boolean mStarted = false;

    public LocationService() {
    }

    @Override
    public void onCreate() {
        EventBus.getDefault().register(this);

        /* Start up the thread running the service.  Note that we create a separate thread because
         * the service normally runs in the process's main thread, which we don't want to block.
         * We also make it background priority so CPU-intensive work will not disrupt our UI.
         */
        HandlerThread thread = new HandlerThread("LocationServiceThread",
                Thread.MIN_PRIORITY);
        thread.start();

        /* Get the HandlerThread's Looper and use it for our Handler */
        mServiceLooper = thread.getLooper();
        mServiceHandler = new Handler(mServiceLooper);

        mServiceHandler.handleMessage(new Message());

        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this.getApplicationContext());
        mLocationRequest = new LocationRequest();
        mLocationRequest.setInterval(1000);
        mLocationRequest.setFastestInterval(1000);
        mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);

        /* Create the Gpx instance */
        mTrackPoints = new ArrayList<>();

        mLocationCallback = new LocationCallback() {
            @Override
            public void onLocationResult(LocationResult locationResult) {
                for (Location location : locationResult.getLocations()) {
                    mServiceHandler.post(() -> {
                        TrackPoint.Builder pointBuilder = new TrackPoint.Builder();
                        pointBuilder.setLatitude(location.getLatitude());
                        pointBuilder.setLongitude(location.getLongitude());
                        pointBuilder.setElevation(location.getAltitude());

                        TrackPoint trackPoint = pointBuilder.build();
                        mTrackPoints.add(trackPoint);
                    });
                }
            }
        };

        startLocationUpdates();
    }

    /**
     * When we stop recording the location events, create a {@link Gpx} object for further
     * serialization. <br>
     * Whatever the outcome of this process, a {@link GpxFileWriteEvent} is emitted in the
     * LocationServiceThread.
     */
    private void createGpx() {
        mServiceHandler.post(() -> {
            ArrayList<TrackSegment> trkSegList = new ArrayList<>();
            trkSegList.add(new TrackSegment(mTrackPoints));

            Track track = new Track(trkSegList, TrekAdvisorContext.APP_FOLDER_NAME + " track");

            ArrayList<Track> trkList = new ArrayList<>();
            trkList.add(track);

            Gpx gpx = new Gpx(trkList, TrekAdvisorContext.APP_FOLDER_NAME, GPX_VERSION);
            try {
                if (!TrekAdvisorContext.DEFAULT_RECORDINGS_DIR.exists()) {
                    TrekAdvisorContext.DEFAULT_RECORDINGS_DIR.mkdir();
                }

                Date date = new Date();
                DateFormat dateFormat = new SimpleDateFormat("dd\\MM\\yyyy-HH:mm:ss", Locale.ENGLISH);
                String gpxFileName = "track-" + dateFormat.format(date) + ".gpx";
                File gpxFile = new File(TrekAdvisorContext.DEFAULT_RECORDINGS_DIR, gpxFileName);
                FileOutputStream fos = new FileOutputStream(gpxFile);
                GPXWriter.INSTANCE.write(gpx, fos);
            } catch (Exception e) {
                // for instance, don't care : we want to stop the service anyway
                // TODO : warn the user that the gpx file could not be saved
            } finally {
                EventBus.getDefault().post(new GpxFileWriteEvent());
            }
        });
    }

    /**
     * Called when the service is started.
     */
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Intent notificationIntent = new Intent(this, MainActivity.class);
        PendingIntent pendingIntent =
                PendingIntent.getActivity(this, 0, notificationIntent, 0);

        Bitmap icon = BitmapFactory.decodeResource(getResources(),
                R.mipmap.ic_launcher);

        Notification.Builder notificationBuilder = new Notification.Builder(getApplicationContext())
                .setContentTitle(getText(R.string.app_name))
                .setContentText(getText(R.string.service_action))
                .setSmallIcon(R.mipmap.ic_launcher)
                .setLargeIcon(Bitmap.createScaledBitmap(icon, 128, 128, false))
                .setContentIntent(pendingIntent)
                .setOngoing(true);

        if (android.os.Build.VERSION.SDK_INT >= 26) {
            /* This is only needed on Devices on Android O and above */
            NotificationManager notificationManager =
                    (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
            NotificationChannel mChannel = new NotificationChannel(NOTIFICATION_ID, getText(R.string.service_description), NotificationManager.IMPORTANCE_DEFAULT);
            mChannel.enableLights(true);
            mChannel.setLightColor(Color.MAGENTA);
            if (notificationManager != null) {
                notificationManager.createNotificationChannel(mChannel);
            }
            notificationBuilder.setChannelId(NOTIFICATION_ID);
        }
        Notification notification = notificationBuilder.build();

        startForeground(SERVICE_ID, notification);

        mStarted = true;
        sendStatus();

        return START_NOT_STICKY;
    }

    @Subscribe
    public void onRecordGpxStartEvent(RecordGpxStartEvent event) {
    }

    /**
     * Create and write a new gpx file. <br>
     * After this is done, a {@link GpxFileWriteEvent} is emitted through event bus so the service
     * can stop properly.
     */
    @Subscribe
    public void onRecordGpxStopEvent(RecordGpxStopEvent event) {
        createGpx();
    }

    /**
     * Self-respond to a {link GpxFileWriteEvent} emitted by the service. <br>
     * When a GPX file has just been written, stop the service and send the status.
     */
    @Subscribe
    public void onGpxFileWriteEvent(GpxFileWriteEvent event) {
        stopSelf();
        mStarted = false;
        sendStatus();
    }

    @Override
    public IBinder onBind(Intent intent) {
        // We don't provide binding, so return null
        return null;
    }

    @Override
    public void onDestroy() {
        stopLocationUpdates();
        mServiceLooper.quitSafely();
        EventBus.getDefault().unregister(this);
    }

    private void startLocationUpdates() {
        if (ActivityCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return;
        }

        mFusedLocationClient.requestLocationUpdates(mLocationRequest,
                mLocationCallback,
                mServiceLooper);
    }

    private void stopLocationUpdates() {
        mFusedLocationClient.removeLocationUpdates(mLocationCallback);
    }

    /**
     * Send the started/stopped boolean status of the service.
     */
    private void sendStatus() {
        EventBus.getDefault().post(new LocationServiceStatus(mStarted));
    }
}

File: app/src/main/java/com/peterlaurence/trekme/service/LocationService.kt
package com.peterlaurence.trekme.service

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.*
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import com.peterlaurence.trekme.MainActivity
import com.peterlaurence.trekme.R
import com.peterlaurence.trekme.core.appName
import com.peterlaurence.trekme.core.track.TrackStatCalculator
import com.peterlaurence.trekme.core.track.TrackStatistics
import com.peterlaurence.trekme.service.event.ChannelTrackPointRequest
import com.peterlaurence.trekme.service.event.GpxFileWriteEvent
import com.peterlaurence.trekme.service.event.LocationServiceStatus
import com.peterlaurence.trekme.ui.events.RecordGpxStopEvent
import com.peterlaurence.trekme.util.gpx.GPXWriter
import com.peterlaurence.trekme.util.gpx.model.Gpx
import com.peterlaurence.trekme.util.gpx.model.Track
import com.peterlaurence.trekme.util.gpx.model.TrackPoint
import com.peterlaurence.trekme.util.gpx.model.TrackSegment
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.io.File
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.*
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

/**
 * A [Started service](https://developer.android.com/guide/components/services.html#CreatingAService)
 * to perform Gpx recordings, even if the user is not interacting with the main application.
 * It is started in the foreground to avoid Android 8.0 (API lvl 26)
 * [background execution limits](https://developer.android.com/about/versions/oreo/background.html).
 * So when there is a Gpx recording, the user can always see it with the icon on the upper left
 * corner of the device.
 *
 * It uses the legacy location API in android.location, not the Google Location Services API, part
 * of Google Play Services. This is because we absolutely need to use only the [LocationManager.GPS_PROVIDER].
 * The fused provider don't give us the hand on that.
 *
 * @author peterLaurence on 17/12/17 -- converted to Kotlin on 20/04/19
 */
class LocationService : Service() {
    private lateinit var serviceLooper: Looper
    private lateinit var serviceHandler: Handler
    private lateinit var locationManager: LocationManager
    private lateinit var locationListener: LocationListener
    private lateinit var recordingsDir: File
    private var locationCounter: Long = 0

    private var trackPoints = mutableListOf<TrackPoint>()
    private lateinit var trackStatCalculator: TrackStatCalculator

    /**
     * A [Channel] to be used for external communication, instead of sharing raw collections across
     * threads (shared mutable state).
     * All operations related to this channel in this class are thread-confined to LocationServiceThread.
     */
    private var channel: Channel<TrackPoint>? = null

    private var mStarted = false

    override fun onCreate() {
        EventBus.getDefault().register(this)

        /* Start up the thread running the service.  Note that we create a separate thread because
         * the service normally runs in the process's main thread, which we don't want to block.
         * We also make it background priority so CPU-intensive work will not disrupt our UI.
         */
        val thread = HandlerThread("LocationServiceThread",
                Thread.MIN_PRIORITY)
        thread.start()

        /* Get the HandlerThread's Looper and use it for our Handler */
        serviceLooper = thread.looper
        serviceHandler = Handler(serviceLooper)


        serviceHandler.handleMessage(Message())

        trackPoints = mutableListOf()

        /* Prepare the stat calculator */
        trackStatCalculator = TrackStatCalculator()

        locationManager = this.getSystemService(Context.LOCATION_SERVICE) as LocationManager
        locationListener = object : LocationListener {
            override fun onLocationChanged(location: Location) {
                locationCounter++
                /* Drop points that aren't precise enough */
                if (location.accuracy > 15) {
                    return
                }

                /* Drop the first 3 points, so the GPS stabilizes */
                if (locationCounter <= 3) {
                    return
                }

                val altitude = if (location.altitude != 0.0) location.altitude else null
                val trackPoint = TrackPoint(location.latitude,
                        location.longitude, altitude, location.time, "")
                trackPoints.add(trackPoint)
                trackStatCalculator.addTrackPoint(trackPoint)
                sendTrackStatistics(trackStatCalculator.getStatistics())
                sendTrackPoint(trackPoint)
            }

            override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {}

            override fun onProviderEnabled(provider: String) {}

            override fun onProviderDisabled(provider: String) {}
        }

        startLocationUpdates()
    }

    /**
     * When we stop recording the location events, create a [Gpx] object for further
     * serialization. <br></br>
     * Whatever the outcome of this process, a [GpxFileWriteEvent] is emitted in the
     * LocationServiceThread.
     */
    private fun createGpx() {
        serviceHandler.post {
            val trkSegList = ArrayList<TrackSegment>()
            trkSegList.add(TrackSegment(trackPoints))

            /* Name the track using the current date */
            val date = Date()
            val dateFormat = SimpleDateFormat("dd-MM-yyyy_HH'h'mm-ss's'", Locale.ENGLISH)
            val trackName = "track-" + dateFormat.format(date)

            val track = Track(trkSegList, trackName)
            track.statistics = trackStatCalculator.getStatistics()

            val trkList = ArrayList<Track>()
            trkList.add(track)

            val wayPoints = ArrayList<TrackPoint>()

            val gpx = Gpx(trkList, wayPoints, appName, GPX_VERSION)
            try {
                val gpxFileName = "$trackName.gpx"
                val gpxFile = File(recordingsDir, gpxFileName)
                val fos = FileOutputStream(gpxFile)
                GPXWriter.write(gpx, fos)
            } catch (e: Exception) {
                // for instance, don't care : we want to stop the service anyway
                // TODO : warn the user that the gpx file could not be saved
            } finally {
                EventBus.getDefault().post(GpxFileWriteEvent())
            }
        }
    }

    /**
     * Called when the service is started.
     */
    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        recordingsDir = intent.getStringExtra(RECORDINGS_PATH_ARG)?.let {
            File(it)
        } ?: error("Recordings dir is mandatory")

        val notificationIntent = Intent(this, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(this, 0, notificationIntent, 0)

        val icon = BitmapFactory.decodeResource(resources,
                R.mipmap.ic_launcher)

        val notificationBuilder = NotificationCompat.Builder(applicationContext, NOTIFICATION_ID)
                .setContentTitle(getText(R.string.app_name))
                .setContentText(getText(R.string.service_location_action))
                .setSmallIcon(R.drawable.ic_my_location_black_24dp)
                .setLargeIcon(Bitmap.createScaledBitmap(icon, 128, 128, false))
                .setContentIntent(pendingIntent)
                .setOngoing(true)

        if (Build.VERSION.SDK_INT >= 26) {
            /* This is only needed on Devices on Android O and above */
            val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            val mChannel = NotificationChannel(NOTIFICATION_ID, getText(R.string.service_location_name), NotificationManager.IMPORTANCE_DEFAULT)
            mChannel.enableLights(true)
            mChannel.lightColor = Color.MAGENTA
            notificationManager.createNotificationChannel(mChannel)
            notificationBuilder.setChannelId(NOTIFICATION_ID)
        }
        val notification = notificationBuilder.build()

        startForeground(SERVICE_ID, notification)

        mStarted = true
        sendStatus()

        return START_NOT_STICKY
    }

    /**
     * Create and write a new gpx file.
     * After this is done, a [GpxFileWriteEvent] is emitted through event bus so the service
     * can stop properly.
     */
    @Subscribe
    fun onRecordGpxStopEvent(event: RecordGpxStopEvent) {
        createGpx()
    }

    /**
     * Self-respond to a {link GpxFileWriteEvent} emitted by the service.
     * When a GPX file has just been written, stop the service and send the status.
     */
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onGpxFileWriteEvent(event: GpxFileWriteEvent) {
        mStarted = false
        sendStatus()
        stopSelf()
    }

    override fun onBind(intent: Intent): IBinder? {
        // We don't provide binding, so return null
        return null
    }

    override fun onDestroy() {
        stopLocationUpdates()
        serviceLooper.quitSafely()
        EventBus.getDefault().unregister(this)
    }

    /**
     * Only use locations from the GPS.
     */
    private fun startLocationUpdates() {
        if (ActivityCompat.checkSelfPermission(this,
                        Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return
        }

        locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 2000, 2f, locationListener, serviceLooper)
    }

    private fun stopLocationUpdates() {
        locationManager.removeUpdates(locationListener)
    }

    /**
     * Send the started/stopped boolean status of the service.
     * Called from main thread.
     */
    private fun sendStatus() {
        EventBus.getDefault().postSticky(LocationServiceStatus(mStarted))
    }

    /**
     * Send updated track statistics.
     * Called from LocationServiceThread.
     */
    private fun sendTrackStatistics(stats: TrackStatistics) {
        EventBus.getDefault().postSticky(stats)
    }

    /**
     * Add a new [TrackPoint] into the channel.
     * Called from LocationServiceThread.
     */
    private fun sendTrackPoint(trackPoint: TrackPoint) {
        channel?.offer(trackPoint)
    }

    @Subscribe
    fun onChannelRequest(event: ChannelTrackPointRequest) = runBlocking {
        GlobalScope.launch {
            val channel = withContext(Dispatchers.Default) {
                newChannel()
            }

            if (channel != null && isStarted) {
                EventBus.getDefault().post(channel)
            }
        }
    }

    /**
     * Creates a new [Channel], filling it with all previously acquired [TrackPoint].
     * This is done in the LocationServiceThread, to ensure thread-safety.
     */
    private suspend fun newChannel(): Channel<TrackPoint>? = suspendCoroutine { cont ->
        serviceHandler.post {
            channel = Channel(capacity = Channel.UNLIMITED)
            trackPoints.forEach {
                channel?.offer(it)
            }
            /* Just in case the service was stopped by the time we get there */
            if (!isStarted) {
                cont.resume(null)
            } else {
                cont.resume(channel)
            }
        }
    }

    companion object {
        const val RECORDINGS_PATH_ARG = "recordingsPath"
        private const val GPX_VERSION = "1.1"
        private const val NOTIFICATION_ID = "peterlaurence.LocationService"
        private const val SERVICE_ID = 126585

        val isStarted: Boolean
            get() {
                val event = EventBus.getDefault().getStickyEvent(LocationServiceStatus::class.java)
                return event?.started ?: false
            }
    }
}
